Implement files versions in Cart Download

juzen2003 opened the pull request:
- Work in progress related to #289 
- Update the results of product_types.json by appending @version to the "product_type" value.
- Update files api. If there is no types parameter specified, all files will be returned. If types parameter is specified, only the specified types & versions will be displayed.
- Pending item:
    - Update download api
    - Add/Fix tests

juzen2003 committed:
- Update the results of product_types.json to have @version appended to
  the original "product_type". (results/views.py, line 1158-1164, 1238-1245)
- Update files api to return only the product_type & version specified
  in types parameter. If there is no types parameter specified, all files
  will be returned. (tools/file_utils.py, line 90-121)

juzen2003 committed:
Fixed the sql query in get_pds_products.

rfrenchseti commented on opus/application/apps/results/views.py:
```diff
@@ -1155,11 +1155,13 @@ def api_get_product_types_for_opus_id(request, opus_id):
-                      'version_number': x[3],
-                      'version_name': x[4]} for x in results]
+    for x in results:
+        prod_type = f'{x[1]}@v{x[4]}' if x[4] != 'Current' else f'{x[1]}@{x[4].lower()}'
```
You could still do this using the list comprehension above. Also I think it would make a little more sense to say:
`f'{x[1]}@current' if x[4] == 'Current' else f'{x[1]}@v{x[4]}`

rfrenchseti commented on opus/application/apps/results/views.py:
```diff
@@ -1233,11 +1235,14 @@ def api_get_product_types_for_search(request):
-                      'version_name': x[4]} for x in results]
+    product_types = []
+    for x in results:
+        prod_type = f'{x[1]}@v{x[4]}' if x[4] != 'Current' else f'{x[1]}@{x[4].lower()}'
```
Same as above

rfrenchseti commented on opus/application/apps/tools/file_utils.py:
```diff
@@ -85,10 +85,40 @@ def get_pds_products(opus_id_list,
     if product_types != ['all']:
-        sql += q('obs_files')+'.'+q('short_name')+' IN %s AND '
-        values.append(product_types)
+        # Because we didn't store @version to the database, when multiple versions (or
```
in the database
product types
are passed in
query the database
in the product_types list
. The query condition will be

rfrenchseti commented on opus/application/apps/tools/file_utils.py:
```diff
@@ -85,10 +85,40 @@ def get_pds_products(opus_id_list,
+
+            # Check and see if the product type has a version specified (look for '@')
+            if '@' in p:
+                idx = p.index('@')
```
Cleaner way to do this:
`prod_type, _, version = p.partition('@')`

rfrenchseti commented on opus/application/apps/tools/file_utils.py:
```diff
@@ -85,10 +85,40 @@ def get_pds_products(opus_id_list,
+                version = p[idx+1:]
+                if version == 'current':
+                    version = 'Current'
+                else:
```
Ideally we would check for "v" here instead of just assuming it

rfrenchseti commented on opus/application/apps/tools/file_utils.py:
```diff
@@ -125,6 +155,10 @@ def get_pds_products(opus_id_list,
+        if version_name != 'Current':
+            short_name = f'{short_name}@v{version_name}'
+        else:
+            short_name = f'{short_name}@{version_name.lower()}'
```
Again cleaner to say:
```
if version_name == 'Current'
    short_name = f'{short_name}@current'
else:
    ...
```

juzen2003 committed:
Restore the behavior of api/product_types.json to be the same as master
branch.

juzen2003 committed:
Update get_pds_products to make sure version modifier like @v1.0, @v1,
@v1.2, @1.0, @1, and @1.2 are all supported.

juzen2003 committed:
Modify api/download api by updating api_create_download. Now download api
will have the following behaviors:
- if no types parameter specified, download all files of the "Current"
  version.
- if a product type is passed in without "@", download all files (all
  versions) of the specified product type.
- if a product type is passed in with "@", download files of the
  specified version & product type.

juzen2003 committed:
Update api_create_download to download all files of all versions if types=all.

juzen2003 committed:
Modify _get_download_info to return product results with different versions.
- In views.py (line 890-1008), update sql query and intelligence to get
  the products with different versions and the corrsponding file sizes.

juzen2003 committed:
Update _get_download_info to create new categories for non "Current"
version files and also check if the files info of a product type should
be added up to total download.
- In views.py (line 1010-1032), update the code to check if the files
  info of a product type should be added up to total download.
- In views.py (line 833-834, 861-874), create new categories for non
  "Current" version files.

juzen2003 committed:
Remove commmented out code in _get_download_info.

juzen2003 committed:
Fixed the download behaviors when multiple versions of the same product
types are selected.
- In cart.html (line 94), use product type + version as the input value
  attribute instead of just the product type. This will make sure we can
  properly select different versions of files in the download pane.
- In file_utils.py (line 130), update the sql query to make sure when
  there is no modifier "@" in types, we will display "Current", this
  will match the behavior of api/download.

juzen2003 committed:
Removed commented out code in get_pds_products.

juzen2003 committed:
Make sure versions other than "Current" are not checked by default in
the cart download pane. Also update the test results for cart api.

juzen2003 committed:
Add tests for _api_cart_download to test on downloading files of different
versions for one opus id.

juzen2003 committed:
Add tests for _api_cart_download to test on downloading files of different
versions for multiple opus ids.

juzen2003 committed:
Update api by adding info regarding to "@" modifier in types parameter

rfrenchseti reviewed (changes requested):
- Version names are different on the Detail tab and the Cart tab. For example, Cassini ISS says "Version: 1" on the Detail tab, but "Cassini ISS-Specific Products V1.0" on the Cart tab.
- Non-current version products are still selected by default on the Cart tab.
- Is there a way to make the version headings on the Cart tab less emphasized? So they're more of a sub-category under the main "Cassini ISS-Specific" heading? This would also serve to group them visually.
- You don't actually need to add files to DropBox to test this, since the shelve files contain all the filenames. Creating a zip file without files present will just create files with zero length.

juzen2003 committed:
Update get_pds_products to return float number version in the result.
This will make sure we have something like 1.0 instead 1 in for the
version in the detail tab.

juzen2003 committed:
- Add non-current versions of the files in cart table view, use the indentation
  and sub-header to identify them as the secondary classes.
- Add "Show previous versions" checkbox to show/display the non-current
  versions of the files in cart table view.
- Add event handler to toggle class "op-cart-ver-hidden" to hide/show
  the non-current versions of the files in cart table view.

juzen2003 committed:
Removed commented out code in _get_download_info.

juzen2003 committed:
- Rename product_cat_list to product_cat_dict.
- Update the function comment about product_cat_dict for _get_download_info.

juzen2003 committed:
Update updateCheckboxes to make sure:
- When select all button is clicked, non-current versions of files won't
  be selected.
- When deselect all button is clicked, non-current versions of files will
  be de-selected.

juzen2003 committed:
Update test_cart_api tests due to the change of product_cat_dict in
_get_download_info.

juzen2003 committed:
Fixed code style in updateCheckboxes in cart.js

juzen2003 committed:
Make sure tooltips are closed when the "v" or "x" buttons are disabled
in download cart pane.

juzen2003 committed:
Update _get_download_info to sort the results based on version number.
This will make sure the download pane has results sorted from the current
to the oldest version.

rfrenchseti commented on opus/application/settings.py:
```diff
@@ -379,3 +379,6 @@
 INVALID_CLASS_CHAR = '~!@$%^&*()+=,./;:"?><[]\{}|`# '
+
+# The symbol is appened after product_type to specify the files version
+FILE_VERSION_MODIFIER = '@'
```
This doesn't really seem like worth including in settings if we're making it a permanent part of the API. We don't do this with anything else in the API.

rfrenchseti commented on opus/application/apps/ui/views.py:
```diff
@@ -636,7 +636,9 @@ def api_init_detail_page(request, **kwargs):
     for version in products:
-        new_products[version] = OrderedDict()
+        # Display "Version: 1.0" instead of "Version: 1" in detail tab
+        pretty_ver = version if version == 'Current' else f'{float(version)}'
```
Instead of using the f'' notation it's more clear just to say str(float(version)) unless you're planning to actually use the f'' formatting in some way.

rfrenchseti commented on opus/application/apps/tools/file_utils.py:
```diff
@@ -87,8 +90,45 @@ def get_pds_products(opus_id_list,
+                else:
+                    # Support cases like @v1.0, @v1, @v1.2, @1.0, @1, and @1.2
+                    if 'v' in version:
+                        version = version[1:]
```
This doesn't necessarily work, since the v could be somewhere else but then we always strip off the first character. How about
version = version.lstrip('v')

rfrenchseti commented on opus/application/apps/tools/file_utils.py:
```diff
@@ -87,8 +90,45 @@ def get_pds_products(opus_id_list,
+        # (obs_files.short_name='coiss_raw' AND obs_files.version_name='Current') OR
+        # ...
+        for i, p in enumerate(product_types):
+            # Uncomment this code if we would like to display all files when the user
```
Is this needed?

rfrenchseti commented on opus/application/apps/tools/file_utils.py:
```diff
@@ -87,8 +90,45 @@ def get_pds_products(opus_id_list,
+                        version = version[1:]
+                    float_ver = float(version)
+                    if float_ver.is_integer():
+                        version = f'{int(float_ver)}'
```
I don't understand this. Why are we converting version to a float then back to an integer string? Isn't that just leaving it the way it started?

rfrenchseti commented on opus/application/apps/tools/file_utils.py:
```diff
@@ -87,8 +90,45 @@ def get_pds_products(opus_id_list,
+                values.append(version)
+            else:
+                # When there is no modifier "@" in types, we will display "Current"
+                # version of the files, this will match the behavior of api/download
```
. This

juzen2003 committed:
Update updateCheckboxes to make sure the non-current version of the files
of a category can be selected when the "v" button of the category is
clicked.

rfrenchseti commented on opus/application/apps/help/api_guide.md:
```diff
@@ -606,10 +606,48 @@ Supported return formats: `zip`.
+calibrated/COISS_2xxx/COISS_2002/data/1460960653_1461048959/N1460973661_1_CALIB.LBL
+calibrated/COISS_2xxx/COISS_2002/data/1462783195_1462915477/N1462840881_1_CALIB.IMG
+calibrated/COISS_2xxx/COISS_2002/data/1462783195_1462915477/N1462840881_1_CALIB.LBL
+calibrated/COISS_2xxx_v1/COISS_2002/data/1460960653_1461048959/N1460973661_1_CALIB.IMG
```
Why are the V1 calibrated files here if we only asked for current and v2?

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -120,13 +120,21 @@ def api_view_cart(request):
-                type['selected'] = 'checked'
+    for name, product_versions in info['product_cat_dict'].items():
+        for ver, types in product_versions.items():
+            for type in types:
```
`type` is the name of a Python built-in function and should not be used as the name of a variable.

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -120,13 +120,21 @@ def api_view_cart(request):
+                    type['selected'] = ''
+                else:
+                    type['selected'] = 'checked'
+    # for name, details in info['product_cat_dict'].items():
```
Remove comments?

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -167,7 +175,7 @@ def api_cart_status(request):
             'total_download_size':        Total size of unique files (bytes)
             'total_download_size_pretty': Total size of unique files (pretty format)
-            'product_cat_list':           List of categories and info:
+            'product_cat_dict':           List of categories and info:
```
The example seems to be wrong now, since it's not showing a dict

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -432,7 +440,7 @@ def api_reset_session(request):
             'total_download_size_pretty': Total size of unique files (pretty format)
-            'product_cat_list':           List of categories and info:
+            'product_cat_dict':           List of categories and info:
                 [
```
Same problem with example

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -540,8 +548,12 @@ def api_create_download(request, opus_id=None, fmt=None):
+        product_types = product_types.lower().split(',')
+    # By default, we want to download all files of the "Current" version if types
+    # parameter is not specified.
+    downloadCurrentOnly = False
```
Don't use mixed-cap variable names in Python

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -540,8 +548,12 @@ def api_create_download(request, opus_id=None, fmt=None):
+    # parameter is not specified.
+    downloadCurrentOnly = False
+    if product_types == []:
+        downloadCurrentOnly = True
```
Easier to write `download_current_only = (product_types == [])`

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -673,62 +685,68 @@ def api_create_download(request, opus_id=None, fmt=None):
+                for file_data in files_version[product_type]:
+                    path = file_data['path']
+                    pretty_name = path.split('/')[-1]
+                    logical_path = path[path.index('/holdings')+9:]
```
I'm not thrilled with hardcoding `/holdings` in the OPUS source code. That's really something that is the domain of PdsFile and OPUS shouldn't know about the directory structure. If we have to do this, let's at least put this string in settings.py.

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -781,22 +799,24 @@ def _get_download_info(product_types, session_id):
-            [
+        'product_cat_dict':           Dict of categories and info:
+            {
              [<Product Type Category>,
```
There should be a key here...

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -812,27 +832,28 @@ def _get_download_info(product_types, session_id):
-    sql += 'AND '+q('obs_files')+'.'+q('version_number')+' >= 900000 '
-    sql += 'ORDER BY '+q('sort')
+    # Put "Current" version on top of others
+    sql += 'ORDER BY '+q('sort')+', FIELD('+q('ver')+' ,"Current") DESC'
```
Can you just say `ver == 'Current'` (and let it sort on True/False) instead of using the `FIELD` function, which is probably slower?

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -965,25 +1011,49 @@ def _get_download_info(product_types, session_id):
+                    p_version = 'current'
+                else:
+                    # Support cases like @v1.0, @v1, @v1.2, @1.0, @1, and @1.2
+                    if 'v' in p_version:
```
p_version = p_version.lstrip('v')

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -965,25 +1011,49 @@ def _get_download_info(product_types, session_id):
+                        p_version = p_version[1:]
+                    float_ver = float(p_version)
+                    if float_ver.is_integer():
+                        p_version = f'{int(float_ver)}'
```
str(int(float_ver))

rfrenchseti reviewed (changes requested):
Some initial comments

juzen2003 committed:
- Update the code to for types parameter
    - Only accept cases like @1, @1.0, @1.2, @current
    - If no version specified, "@" is not seen in the types, will return
      the current version
    - If we passed in something like "coiss_calib@", it means the version
      is an empty string for that specific product type, will return nothing for that product type.
    - If "types=" is passed in, we treat it as an empty string, will return
      nothing
- Treat "1" and "1.0" differently:
    - In detail tab:
        - If version is 1 (store "1" in db), Version: 1 is displayed
        - If version is 1.0 (store "1.0" in db), Version: 1.0 is displayed
    - In cart download pane:
        - If version is 1 (store "1" in db), Version: 1 is displayed
        - If version is 1.0 (store "1.0" in db), Version: 1.0 is displayed

juzen2003 committed:
Add fixes based on github comments.

juzen2003 committed:
Fixed tests in test_cart_api.py

juzen2003 committed:
Add tests to test against unmatched version name "1.0" & "v1" in test_cart_api.py

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -673,62 +682,68 @@ def api_create_download(request, opus_id=None, fmt=None):
+                for file_data in files_version[product_type]:
+                    path = file_data['path']
+                    pretty_name = path.split('/')[-1]
+                    logical_path = path[path.index(settings.PDS_HOLDINGS_DIR)+9:]
```
The 9 should be len(settings.PDS_HOLDINGS_DIR)

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -673,62 +682,68 @@ def api_create_download(request, opus_id=None, fmt=None):
+                    checksum = file_data['checksum']
+                    size = file_data['size']
+                    pretty_name = path.split('/')[-1]
+                    logical_path = path[path.index(settings.PDS_HOLDINGS_DIR)+9:]
```
The 9 should be len(settings.PDS_HOLDINGS_DIR)

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -845,11 +864,19 @@ def _get_download_info(product_types, session_id):
+            product_cat_dict[pretty_name] = {}
+            product_cat_dict[pretty_name][ver] = cur_product_list
+        else:
+            try:
```
This is not a good use of try/except because the KeyError could be from either `[pretty_name]` or [`ver`]. Better to just check using `in` explicitly.

rfrenchseti commented on opus/application/apps/help/api_guide.md:
```diff
@@ -877,6 +885,45 @@ Examples:
 }
 %ENDCODE%
 
+* Retrieve raw images ("Current" version) and calibrated images (version 1.0) for a Cassini ISS observation in JSON format.
```
This should say "version 1"

rfrenchseti commented on opus/application/apps/help/api_guide.md:
```diff
@@ -836,7 +852,7 @@ Examples:
 %ENDCODE%
 
-* Retrieve raw images only for a Galileo SSI observation in JSON format.
+* Retrieve raw images ("Current" version) only for a Galileo SSI observation in JSON format.
```
This is pointing out that we are losing functionality. There is no longer a way to ask for "all" versions of a product type. It used to be when you specified a product type, you got all the versions, but now you only get the "current" version. Perhaps it would be worth adding an "@all" feature to the `files` API?

rfrenchseti reviewed (changes requested):
- "Select All Product Types" disables all of the checkmarks at the top of sections, even for the sub-versions. But in those cases the product types weren't actually selected so it's the wrong thing to do.
- Changing any of the "current" version checkboxes causes the total size and number of files  at the top to go to zero. Only selecting a sub-version gives a good total size and number of files.

juzen2003 committed:
Make sure the non-current version check marks are not disabled when
"Select all product types" in download pane is clicked.

juzen2003 committed:
Fixed the issue of total files size & count not being updated when selecting
the current version of files in the download cart pane. This is done by
fixing the comparison of version names in _get_download_info when calculating
the total files size & count.

juzen2003 committed:
Add "@all" to return all versions of files for a product type (line 103-
111 in get_pds_products).

juzen2003 committed:
Add an example of "@all" in the api guide.

juzen2003 committed:
Fixed the issue of displaying the wrong total files size & count when the
download page is initialized. Use default_checked in the db to determine
if the files should be counted.

juzen2003 committed:
Fixed tests in test_cart_api.py and add tests for "@all"

juzen2003 committed:
Update comments related to sql in _get_download_info.

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -253,16 +260,31 @@ var o_cart = {
+        if ($(target).hasClass("op-cart-select-all-btn")) {
+            // When "Select all product types" is clicked, only select the current version
+            // of the files
+            for(let input of checkboxesOptions) {
```
Add space after `for`

rfrenchseti commented on opus/application/static_media/css/opus.css:
```diff
@@ -1485,6 +1485,17 @@ panel */
 }
 
+.op-cart-ver-sub-header {
+    /* font-size: 0.8rem; */
```
Remove comment?

rfrenchseti commented on opus/application/test_api/test_cart_api.py:
```diff
@@ -311,7 +311,7 @@ def test__api_cart_add_download_one(self):
         self._run_json_equal(url, expected)
         url = '/__cart/add.json?opusid=co-iss-n1460960653&download=1&reqno=101010101'
-        expected = {'total_download_count': 23, 'total_download_size': 24063890, 'total_download_size_pretty': '22M', 'product_cat_list': [['Cassini ISS-Specific Products', [{'slug_name': 'coiss_raw', 'tooltip': 'Raw image files (*.IMG) for Cassini ISS. Images are in VICAR format, uncalibrated, and in units of DN (data number). Associated labels (*.LBL) are text files that contain information abo ...
+        expected = {'total_download_count': 7, 'total_download_size': 5525793, 'total_download_size_pretty': '5M', 'product_cat_dict': {'Cassini ISS-Specific Products': {'Current': [{'slug_name': 'coiss_raw', 'tooltip': 'Raw image files (*.IMG) for Cassini ISS. Images are in VICAR format, uncalibrated, and in units of DN (data number). Associated labels (*.LBL) are text files that contain informa ...
```
I just noticed there's an entire class of tests missing that you could add now. All of the cart operators like add, remove, and reset take an optional types= parameter when download=1 is specified, but we never test this. This would be a good time to add a few tests that add that parameter and also check versions at the same time.

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -965,25 +1007,41 @@ def _get_download_info(product_types, session_id):
+            elif short_name == p:
+                is_adding_up_to_total = True
+                break
+        if (product_types == ['all'] and checked) or is_adding_up_to_total:
```
This doesn't seem right to me. _get_download_info should not be checking whether an item is checked by default. "all" should really mean all. Perhaps we need a function flag that says something like "count_default_checked_only"? Or is there a way to pass in the list of default-checked product types?

rfrenchseti replied:
I thought about this some more. Perhaps this is OK after all, since it's only for internal use. But there needs to be a comment added to the function header saying that "all" doesn't really mean ALL.

rfrenchseti replied:
I think what you want to say is:
The resulting totals are limited to the given product_types. ['all'] means include all product types that are "checked by default" in the database.

juzen2003 committed:
Update based on github feedback and add tests in test_cart_api.py

juzen2003 committed:
- Update "Select all current version product types" & "Deselect all product
  types" behaviors:
    - When all current version inputs (non-current version inputs are not
      included) are checked, "Select all current version product types" is disabled
    - When all inputs (both current & non-current version inputs are included)
      are unchecked, "Deselect all product types" is disabled.
- Update "v" & "x" behaviors for each category:
    - When all inputs in a category (not including the non-current version
      inputs) are checked, "v" of that category will be disabled. Non-current
      version inputs are under their own category, Version: x.
    - When all inputs in a category (not including the non-current version
      inputs) are unchecked, "x" of that category will be disabled. Non-current
      version inputs are under their own category, Version: x.

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -142,8 +142,10 @@ var o_cart = {
             let isAllCatOptionsUnchecked = o_cart.isAllOptionStatusTheSame(productInputs, false);
 
-            let allCheckboxesOptions = $(".op-download-options-product-types input");
+            // let allCheckboxesOptions = $(".op-download-options-product-types input");
```
Remove comment?

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -175,14 +179,21 @@ var o_cart = {
             o_cart.updateCheckboxes(e.currentTarget, productList);
 
-            let allCheckboxesOptions = $(".op-download-options-product-types input");
+            // let allCheckboxesOptions = $(".op-download-options-product-types input");
```
Remove comment?

rfrenchseti commented on opus/application/test_api/test_cart_api.py:
```diff
@@ -2163,6 +2202,27 @@ def test__api_cart_recyclebin_reset(self):
+        url = '/__cart/status.json?reqno=456'
+        expected = {'recycled_count': 0, 'count': 1, 'reqno': 456}
+        self._run_json_equal(url, expected)
+        url = '/__cart/reset.json?reqno=42&recyclebin=0'
```
Maybe added types= here too?

rfrenchseti commented on opus/application/test_api/test_cart_api.py:
```diff
@@ -778,7 +802,22 @@ def test__api_cart_remove_one_download(self):
+        expected = {'recycled_count': 0, 'count': 0, 'reqno': 789}
+        self._run_json_equal(url, expected)
+
+    def test__api_cart_remove_download_ver(self):
```
This is sort of a boring test, since the cart is empty after the remove, so the list of product types doesn't matter. Is there another case you can do where something is left in the cart after the remove?

juzen2003 committed:
- Removed commented out code in cart.js.
- Update comment for _get_download_info in views.py.
- Update tests for test__api_cart_remove_download_ver in test_cart_api.py.
